<a href="https://colab.research.google.com/github/AlejandroC-25/Ciencia-y-analitica-de-datos-A00835060-/blob/main/Actividad3BDManipulacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Alejandro Cena Peart
*   MATRÍCULA: A00835060


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [9]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [10]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [11]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [12]:
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


**Se encontraron 5 películas con duración mayor a 5 horas. La duración (runtime) está registrada en minutos, por lo que el filtro usado fue más de 300 minutos.**

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [13]:
pd.read_sql(sqla.text("""
    SELECT COLUMN_NAME, COLUMN_COMMENT
    FROM information_schema.COLUMNS
    WHERE TABLE_SCHEMA = 'Netflix'
      AND TABLE_NAME = 'movie'
      AND COLUMN_NAME = 'available_globally'
"""), conn).iloc[0, 1]

'A flag that indicates if this title is available outside US'

In [14]:
query = sqla.text("""
    SELECT ROUND(100 * SUM(available_globally = 0) / COUNT(*), 2) AS pct_solo_eu
    FROM movie
    WHERE available_globally IS NOT NULL
""")
pd.read_sql(query, conn)

,pct_solo_eu
0,83.57


**El 83.57 % de las películas están disponibles únicamente en EE. UU. (available_globally = 0, es decir, no disponibles fuera de EE. UU.).**

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [15]:
query = sqla.text("""
    SELECT locale, COUNT(*) AS n_peliculas
    FROM movie
    WHERE locale IS NOT NULL
    GROUP BY locale
    ORDER BY n_peliculas DESC
""")
pd.read_sql(query, conn)

,locale,n_peliculas
0,en,532


In [16]:
query = sqla.text("""
    SELECT COUNT(*) AS n_locale_null
    FROM movie
    WHERE locale IS NULL
""")
pd.read_sql(query, conn)

,n_locale_null
0,11255



**El único idioma/región original registrado en locale es en inglés, con 532 películas. Hay 11,255 registros con locale NULL**

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [17]:
query = sqla.text("""
    SELECT COUNT(*) AS n_titulos_diferentes
    FROM movie
    WHERE title <> original_title
""")
pd.read_sql(query, conn)

,n_titulos_diferentes
0,3947


In [18]:
query = sqla.text("""
    SELECT
        CASE WHEN locale IS NULL THEN 'locale NULL' ELSE locale END AS locale_valor,
        CASE
            WHEN original_title IS NULL THEN 'original_title NULL'
            WHEN title <> original_title THEN 'titulos diferentes'
            ELSE 'titulos iguales'
        END AS relacion_titulos,
        COUNT(*) AS n
    FROM movie
    GROUP BY locale_valor, relacion_titulos
    ORDER BY locale_valor, relacion_titulos
""")
pd.read_sql(query, conn)

,locale_valor,relacion_titulos,n
0,en,original_title NULL,532
1,locale NULL,original_title NULL,7291
2,locale NULL,titulos diferentes,3947
3,locale NULL,titulos iguales,17


**Hay 3,947 películas con title diferente de original_title, y todas tienen locale NULL. Los resultados no coinciden: hay 11,255 registros con locale NULL, pero solo 3,947 tienen títulos diferentes. Los 7,308 restantes se dividen en 7,291 que no tienen original_title (no se pueden comparar) y 17 que tienen el mismo título en ambos campos. Además, las 532 películas con locale = 'en' tampoco tienen original_title. Entre las películas que sí se pueden comparar (3,964), el 99.6 % tiene títulos diferentes, lo que apoya la suposición en ese grupo; sin embargo, solo el 35.2 % de los registros con locale NULL se puede comparar. Conclusión: la suposición no es válida de forma general. Aunque en los casos comparables los títulos casi siempre son diferentes, para la mayoría de los registros no hay datos para confirmarlo, por lo que es más prudente interpretar locale NULL como "idioma no registrado" y no como "otro idioma".**



5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [19]:
query = sqla.text("""
    SELECT m.id, m.title, MAX(v.cumulative_weeks_in_top10) AS semanas_top10
    FROM movie m
    JOIN view_summary v ON v.movie_id = m.id
    GROUP BY m.id, m.title
    HAVING MAX(v.cumulative_weeks_in_top10) = (
        SELECT MAX(cumulative_weeks_in_top10)
        FROM view_summary
        WHERE movie_id IS NOT NULL
    )
""")
pd.read_sql(query, conn)

,id,title,semanas_top10
0,11197,The Boss Baby,22


**La película que más tiempo ha permanecido en el top 10 es The Boss Baby, con 22 semanas acumuladas.**

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [20]:
query = sqla.text("""
    SELECT id, title, n_temporadas
    FROM (
        SELECT t.id, t.title, COUNT(s.id) AS n_temporadas,
               RANK() OVER (ORDER BY COUNT(s.id) DESC) AS posicion
        FROM tv_show t
        JOIN season s ON s.tv_show_id = t.id
        GROUP BY t.id, t.title
    ) ranking
    WHERE posicion <= 5
    ORDER BY n_temporadas DESC, title
""")
pd.read_sql(query, conn)

,id,title,n_temporadas
0,347,Grey's Anatomy,20
1,1020,Naruto Shippuden,20
2,564,Heartland (2007),17
3,2207,It's Always Sunny in Philadelphia,16
4,431,NCIS,15
5,4326,Raw,15
6,881,Supernatural (2005),15


**Los programas con mayor cantidad de temporadas son Grey's Anatomy y Naruto Shippuden, 20 cada uno, Heartland, 17 temporadas, It's Always Sunny in Philadelphia con 16 temporadas y, empatados en quinto lugar con 15 temporadas, NCIS, Raw y Supernatural. Por eso el resultado incluye 7 programas.**

**El número de temporadas se calculó contando los registros de season de cada programa (tv_show_id), y se usó RANK() para no dejar fuera programas con mismo numero de temporadas**

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [21]:
query = sqla.text("""
    SELECT DISTINCT start_date, end_date
    FROM view_summary
    WHERE duration = 'SEMI_ANNUALLY'
    ORDER BY start_date
""")
pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


**Los resúmenes con duration = 'SEMI_ANNUALLY' abarcan dos intervalos: del 1 de julio al 31 de diciembre de 2023 (segundo semestre de 2023) y del 1 de enero al 30 de junio de 2024 (primer semestre de 2024).**

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [22]:
query = sqla.text("""
    SELECT s.season_number, s.title, v.views
    FROM tv_show t
    JOIN season s ON s.tv_show_id = t.id
    JOIN view_summary v ON v.season_id = s.id
    WHERE t.title = 'Grey''s Anatomy'
      AND v.duration = 'SEMI_ANNUALLY'
      AND v.start_date = '2024-01-01'
      AND v.end_date = '2024-06-30'
    ORDER BY v.views DESC, s.season_number
""")
resultado_sql = pd.read_sql(query, conn)
resultado_sql

,season_number,title,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


**En el primer semestre de 2024 (2024-01-01 a 2024-06-30) las 20 temporadas de Grey's Anatomy tienen registro en view_summary. La temporada 1 encabeza con 3.6 M de vistas, seguida de la temporada 2 (3.1 M), y a partir de ahí las vistas descienden de forma casi monótona hasta 1.7 M en las temporadas 17 y 18. Hay dos excepciones: la temporada 19 (2.0 M) supera a las temporadas 15 a 18, y la temporada 20 cae a 100,000 vistas.**

**las primeras temporadas son las más vistas y las vistas bajan poco a poco conforme avanza la serie. Esto sugiere que mucha gente empieza a verla desde el principio y que cada vez menos personas llegan a las temporadas más avanzadas. La temporada 19 rompe un poco esa tendencia con 2.0 M de vistas supera a las temporadas 15 a 18, lo que podría indicar más interés en una temporada reciente. La temporada 20 tiene muy pocas vistas, posiblemente porque estuvo disponible solo una parte del periodo o se estrenó hace poco. Estas explicaciones son hipótesis, ya que los datos no son suficientes para confirmarlas.**

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [23]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [24]:
grey = tv_show.loc[tv_show['title'] == "Grey's Anatomy", ['id']].rename(columns={'id': 'tv_show_id'})
temporadas = season[['id', 'tv_show_id', 'season_number', 'title']] \
    .rename(columns={'id': 'season_id'}) \
    .astype({'season_number': 'Int64'})

vistas = view_summary[
    (view_summary['duration'] == 'SEMI_ANNUALLY')
    & (pd.to_datetime(view_summary['start_date']) == '2024-01-01')
    & (pd.to_datetime(view_summary['end_date']) == '2024-06-30')
][['season_id', 'views']]

resultado = (
    grey.merge(temporadas, on='tv_show_id')
        .merge(vistas, on='season_id')
        .sort_values(['views', 'season_number'], ascending=[False, True])
        [['season_number', 'title', 'views']]
        .reset_index(drop=True)
)
resultado

,season_number,title,views
0,1,Grey's Anatomy: Season 1,3600000
1,2,Grey's Anatomy: Season 2,3100000
2,3,Grey's Anatomy: Season 3,2900000
3,4,Grey's Anatomy: Season 4,2900000
4,5,Grey's Anatomy: Season 5,2900000
5,6,Grey's Anatomy: Season 6,2800000
6,7,Grey's Anatomy: Season 7,2700000
7,8,Grey's Anatomy: Season 8,2600000
8,9,Grey's Anatomy: Season 9,2500000
9,10,Grey's Anatomy: Season 10,2400000


**Se replicó la consulta de la pregunta 8 con pandas, el resultado coincide con el de la pregunta 8:**

**mismas 20 temporadas, mismas vistas y mismo orden, lo que muestra que la lógica relacional de SQL se puede reproducir utilizando pandas.**

`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [25]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [26]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [27]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Actividad 3-CAD"
os.chdir(DIR)

In [28]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [29]:
docs = db.collection('EMPLOYEE').stream()
df_employee = pd.DataFrame([{'id': doc.id, **doc.to_dict()} for doc in docs])
df_employee

,id,emp_fname,emp_lname,emp_hiredate
0,8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
1,Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
2,lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
3,yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [30]:
firebase_admin.delete_app(firebase_admin.get_app())

**Para leer la colección EMPLOYEE de Firestore se usa db.collection('EMPLOYEE').stream(), que recorre todos los documentos de la colección. Cada documento trae sus datos con to_dict(), pero el identificador (id) no viene incluido ahí, sino que se obtiene aparte con doc.id. Por eso, para cada documento se junta el id con sus datos y con todo eso se arma el DataFrame con pd.DataFrame. El resultado muestra 4 empleados con las columnas id, emp_fname, emp_lname y emp_hiredate.**

---

**Declaración de uso de IA**

*   Anthropic. (2026). *claude sonnet 5.5* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://claude.ai/new

*   Google. (2026). *gemini* [Modelo de lenguaje grande], utilizado para resolver dudas de conceptos y mejorar la redacción en las repsuestas 8 y 4. https://gemini.google.com/app?hl=es-MX


---